Document --show-hidden-metrics-for-version flag (#17471)
kubernetes/website: content/en/docs/concepts/cluster-administration/monitoring.md

---
title: Monitoring Control Plane Components
reviewers:
- brancz
- logicalhan
- RainbowMango
content_template: templates/concept
weight: 60
---

{{% capture overview %}}

System component metrics can give a better look into what is happening inside them. Metrics are particularly useful for building dashboards and alerts.

{{% /capture %}}

{{% capture body %}}

## Metrics in Kubernetes

> Kubernetes control plane components use prometheus as the default client for exposing metrics. In most cases those metrics are available on `/metrics` endpoint of the HTTP server. For components that doesn't expose endpoint by default it can be enabled using `--bind-address` flag.

Reviewer: It doesn't really matter to Kubernetes what client you use to scrape these metrics.

```suggestion
Kubernetes control plane components publish metrics in Prometheus text format. In most cases those metrics are available on `/metrics` endpoint of the HTTP server. For components that doesn't expose endpoint by default it can be enabled using `--bind-address` flag.
```

At the end of this page, you can add:
```markdown
{{% capture whatsnext %}}
* Read about the [Prometheus text format](https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md#text-based-format) for metrics
{{% /capture %}}
```

Examples of those components:

> * kube-controller-manager

Reviewer: ```suggestion
* {{< glossary_tooltip term_id="kube-controller-manager" text="kube-controller-manager" >}}
```
etc

* kube-proxy
* kube-apiserver
* kube-scheduler
* kubelet (metrics exposed on `/metrics/cadvisor` and `/metrics/resource` do not have same guarantee)

> For clusters with RBAC enabled accessing metrics requires authorization through service account with `ClusterRole` for `/metrics` url.

Reviewer: I think this is better:
```suggestion
If your cluster uses {{< glossary_tooltip term_id="rbac" text="RBAC" >}}, reading metrics requires authorization via a user, group or ServiceAccount with a ClusterRole that allows accessing `/metrics`.
```

@serathius do you agree?

> 

Reviewer: ```suggestion
For example:
```

or consider using a `codenew` block so that the ClusterRole is downloadable.

```
apiVersion: rbac.authorization.k8s.io/v1	
kind: ClusterRole	
metadata:	
  name: prometheus	
rules:	
  - nonResourceURLs:	
      - "/metrics"	
    verbs:	
      - get
```

> ## Metric Lifecycle 

Reviewer: nit:
```suggestion
## Metric lifecycle
```

> Alpha metric -> Stable metric -> Deprecated metric -> Hidden metric -> Deletion

Reviewer: nit:
```suggestion
Alpha metric →  Stable metric →  Deprecated metric →  Hidden metric → Deletion
```
This might work better as a diagram?

Alpha metrics have no stability guarantees; as such they can be modified or deleted at any time.

> Stable metrics can be guaranteed to not change, except that the metric may become marked deprecated for a future Kubernetes version. By not change, we mean three things:

Reviewer: ```suggestion
Stable metrics can be guaranteed to not change, except that the metric may become marked deprecated for a future Kubernetes version. Specifically, stability means:
```
(from [avoid using “we”](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we))

* the metric itself will not be deleted (or renamed)
* the type of metric will not be modified
* no labels can be added or removed from this metric

> List of currently supported stable metrics can be found [here](https://github.com/kubernetes/kubernetes/blob/master/test/instrumentation/testdata/stable-metrics-list.yaml)

Reviewer: Move this to a what's next section.
For example:
```markdown
{{% capture whatsnext %}}
* See the list of [stable Kubernetes metrics](https://github.com/kubernetes/kubernetes/blob/master/test/instrumentation/testdata/stable-metrics-list.yaml)
{{% /capture %}}
```

Deprecated metric are annotated with a Kubernetes version, from which point that metric will be considered deprecated. When a stable metric undergoes the deprecation process, we are signaling that the metric will eventually be deleted. 

> 

Reviewer: Maybe link to and contrast https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior ?

(possibly with a mention in the page's _What's next_ section?)

Before deprecation:

```
# HELP some_counter this counts things
# TYPE some_counter counter
some_counter 0
```

After deprecation:

```
# HELP some_counter (Deprecated from 1.15) this counts things
# TYPE some_counter counter
some_counter 0
```

> Hidden metrics will no longer be exposed by default and will require manual component configuration by cluster administrators;

Reviewer: ```suggestion
Hidden metrics will no longer be exposed by default; to use a hidden metric, you need to override the configuration for the relevant cluster component.
```
(from [address the reader as “you”](https://kubernetes.io/docs/contribute/style/style-guide/#address-the-reader-as-you))

Note that this changes the grammar; if you adopt this suggestion, other related lines will need to end with periods rather than semicolons.
